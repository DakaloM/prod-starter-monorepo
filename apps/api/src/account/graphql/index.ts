
export * from './user';
export * from './users';
export * from './password';
export * from './createUser';
export * from './updateUser';
// export * from './userRegister'
export * from './hasUser'
