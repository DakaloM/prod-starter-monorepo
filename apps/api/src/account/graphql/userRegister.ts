// import { allow } from 'graphql-shield';
// import { builder } from '~/graphql/builder';
// import { UserSchema } from './user';
// import { userRegister } from '../mutations';
// import { TokensRef } from '~/auth/graphql/tokens';
// import { isActiveUser } from '../shield';

// const UserRegisterInput = builder.inputType('UserRegisterInput', {
//   fields: (t) => ({
//     name: t.string({ required: true }),
//     surname: t.string({ required: true }),
//     email: t.string({ required: true }),
//     idNumber: t.string({required: true}),
//     password: t.string({required: true}),
//     confirmPassword: t.string({required: true}),
//     mobileNumber: t.string({required: true}),
//     businessNumber: t.string({required: false}),
//     privateNumber: t.string({required: false}),
//   }),
// });

// builder.mutationField('userRegister', (t) => (
//   t.field({
//     shield: allow,
//     description: 'Register a new user',
//     args: {
//       input: t.arg({type: UserRegisterInput, required: true}),
//     },
//     type: UserSchema,
//     resolve: async (_root, args, ctx) => {
//       return userRegister(args.input, ctx);
//     }
//   })
// ))
