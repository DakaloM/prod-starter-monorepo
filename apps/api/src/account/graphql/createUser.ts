
import { builder } from "~/graphql/builder";
import { GenderSchema, RaceSchema, TitleSchema, UserRoleSchema, UserSchema, UserStatusSchema } from "./user";
import { createUser } from "../mutations";
import { isDashboardOperator } from "../shield";

const CreateUserInput = builder.inputType('CreateUserInput', {
  fields: (t) => ({
    name: t.string({required: true}),
    middleName: t.string({required: false}),
    surname: t.string({required: true}),
    email: t.string({required: true}),
    title: t.field({type: TitleSchema, required: true}),
    race: t.field({type: RaceSchema, required: true}),
    gender: t.field({type: GenderSchema, required: true}),
    birthDate: t.field({type: 'Date', required: true}),
    idNumber: t.string({required: true}),
    role: t.field({type: UserRoleSchema, required: true}),
    status: t.field({type: UserStatusSchema})
  })
});

builder.mutationField('createUser', (t) => (
  t.field({
    shield: isDashboardOperator,
    description: 'Create a user',
    args: {
      input: t.arg({type: CreateUserInput, required: true}),
    },
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      return await createUser(args.input, ctx);
    }
  })
))