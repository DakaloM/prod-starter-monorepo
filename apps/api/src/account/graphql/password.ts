import { builder } from '~/graphql/builder';

import { forgotPassword, recoverPassword } from '~/account/mutations';
import { TokensRef } from '~/auth/graphql/tokens';
import { isAuthClient } from '~/auth/shield';

builder.mutationField('forgotPassword', (t) =>
  t.field({
    shield: isAuthClient,
    description: 'Initiate forgot password flow',
    args: {
      email: t.arg({ type: 'EmailAddress', required: true }),
    },
    type: 'String',
    resolve: async (_root, args, ctx) => {
      return forgotPassword(args, ctx);
    },
  }),
);

builder.mutationField('recoverPassword', (t) =>
  t.field({
    shield: isAuthClient,
    description: 'Complete forgot password flow',
    args: {
      email: t.arg({ type: 'EmailAddress', required: true }),
      code: t.arg.string({ required: true }),
      password: t.arg.string({ required: true }),
    },
    type: TokensRef,
    resolve: async (_root, args, ctx) => {
      return recoverPassword(args, ctx);
    },
  }),
);
