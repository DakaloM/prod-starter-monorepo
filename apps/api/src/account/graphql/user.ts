import { builder } from '~/graphql/builder';
import { isActiveUser, isDashboardOperator } from '~/account/shield';
import { Gender, UserRole, UserStatus, Race, Title, User } from '~/account/user';
import { nullable } from 'zod';
import { getUserById } from '../queries';
import { allow, or } from 'graphql-shield';


export const UserRoleSchema = builder.enumType(UserRole, {
  name: 'UserRole',
  description: 'User role',
});

export const getEnumValues = <T extends Record<string, any>>(enumType: T) => {
  return Object.keys(enumType).reduce((acc, key) => {
    return {
      ...acc,
      [key]: { value: enumType[key] },
    };
  }, {});
};


export const GenderSchema = builder.enumType(Gender, {
  name: 'Gender',
  description: 'Gender',
});

export const RaceSchema = builder.enumType(Race, {
  name: 'Race',
  description: 'Race',
});

export const TitleSchema = builder.enumType(Title, {
  name: 'Title',
  description: 'Title',
});

export const UserStatusSchema = builder.enumType(UserStatus, {
  name: 'UserStatus',
  description: 'User status',
});

export const UserSchema = builder.objectType(User, {
  name: 'User',
  shield: allow,
  description: 'A user is an human user or bot that can interact with the system.',

  fields: (t) => ({
    id: t.exposeID('id'),
    email: t.field({
      type: 'EmailAddress',
      resolve: (root) => root.email,
    }),
    role: t.expose('role', { type: UserRoleSchema, description: 'The role of the user.' }),
    name: t.exposeString('name'),
    surname: t.exposeString('surname'),
    middleName: t.exposeString('middleName',{nullable: true}),
    title: t.exposeString('title' ,{nullable: true}),
    gender: t.exposeString('gender' ,{nullable: true}),
    idNumber: t.exposeString('idNumber'),
    race: t.exposeString('race' ,{nullable: true}),
    status: t.exposeString('status'),
    birthDate: t.expose('birthDate', {type: 'Date', nullable: true}),
    sequence: t.exposeInt('sequence'),
    applicantId: t.exposeString('applicantId', {nullable: true})
   
  }),
  
});

builder.queryField('user', (t) =>
  t.field({
    shield: or(isActiveUser, isDashboardOperator),
    description: 'A user',
    args: {
      id: t.arg.id({ required: true }),
    },
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      return getUserById(args.id.toString(), ctx);
    },
  }),
);

builder.queryField('me', (t) =>
  t.field({
    shield: isActiveUser,
    description: 'A user',
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      const id = ctx.auth.user?.id.toString() || '';

      return getUserById(id, ctx);
    },
  }),
);




