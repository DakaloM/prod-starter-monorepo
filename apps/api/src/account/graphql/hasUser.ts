import { builder } from "~/graphql/builder";
import { UserSchema } from "./user";
import { getUserById } from "../queries";

//@ts-ignore
export const HasUserSchema = builder.interfaceType('HasUser', {
 
  fields: (t) => ({
    user: t.field({
      type: UserSchema,
      args: {},
      description: 'User object',
      
      resolve: async (root, args, ctx) => {
        //@ts-ignore
        return await getUserById(root.userId, ctx);
      },
    }),
  }),
});
