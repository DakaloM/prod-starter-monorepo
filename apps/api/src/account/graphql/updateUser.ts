
import { builder } from "~/graphql/builder";
import { GenderSchema, RaceSchema, TitleSchema, UserRoleSchema, UserSchema, UserStatusSchema } from "./user";
import { isActiveUser } from "../shield";
import { updateUser } from "../mutations";


const UpdateUserInput = builder.inputType('UpdateUserInput', {
  fields: (t) => ({
    id: t.string({required: true}),
    name: t.string({required: false}),
    middleName: t.string({required: false}),
    surname: t.string({required: false}),
    email: t.string({required: false}),
    title: t.field({type: TitleSchema, required: false}),
    race: t.field({type: RaceSchema, required: false}),
    gender: t.field({type: GenderSchema, required: false}),
    birthDate: t.field({type: 'Date', required: false}),
    idNumber: t.string({required: false}),
    role: t.field({type: UserRoleSchema, required: false}),
    status: t.field({type: UserStatusSchema, required: false})
  })
});

builder.mutationField('updateUser', (t) => (
  t.field({
    shield: isActiveUser,
    description: 'Update user',
    args: {
      input: t.arg({type: UpdateUserInput, required: true}),
    },
    type: UserSchema,
    resolve: async (_root, args, ctx) => {
      return updateUser(args.input, ctx);
    }
  })
))