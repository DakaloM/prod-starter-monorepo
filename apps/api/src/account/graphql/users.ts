import { builder } from '~/graphql/builder';

import { createPayload } from '~/domain/search';

import { getUsers } from '../queries';
import { isActiveUser } from '../shield';
import { UserRoleSchema, UserSchema, getEnumValues } from './user';
import { UserRole } from '../user';
import { EnumRef, ValuesFromEnum } from '@pothos/core';

export const RoleFilterSchema = builder.enumType('RoleFilter', {
  values: {
    ...getEnumValues(UserRole),
  },
}) as unknown as EnumRef<ValuesFromEnum<any>, ValuesFromEnum<any>>;

export const UsersFilter = builder.inputType('UsersFilter', {
  fields: (f) => ({
    role: f.field({ type: RoleFilterSchema, required: false }),
    name: f.field({ type: 'String', required: false }),
    notRole: f.field({ type: RoleFilterSchema, required: false }),
  }),
});



const UsersPayloadSchema = createPayload('UsersPayload', UserSchema);

builder.queryField('users', (t) =>
  t.field({
    shield: isActiveUser,
    description: 'List of users',
    args: {
      limit: t.arg.int({ defaultValue: 20 }),
      page: t.arg.int({ defaultValue: 1 }),
      filter: t.arg({ required: false, type: UsersFilter }),
    },
    type: UsersPayloadSchema,
    resolve: async (_root, args, ctx) =>
      getUsers(
        {
          limit: args.limit || 20,
          page: args.page || 1,
          ...args.filter,
        },
        ctx,
      ),
  }),
);

interface Count {
  count: string;
}
