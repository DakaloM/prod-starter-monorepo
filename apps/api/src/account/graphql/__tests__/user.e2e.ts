import { UserRole } from '@imax/client';
import { faker } from '@imax/testkit';

import crypto from 'node:crypto';
import { updateUser } from '~/account/mutations';
import { User, UserStatus } from '~/account/user';
import { Context } from '~/context';
import { createContext, destroyContext } from '~/test/context';
import { Environment, createEnv, destroyEnv } from '~/test/environment';
import { Factory, createFactory } from '~/test/factory';
import { gql, request } from '~/test/graphql';

describe('account/graphql/user', () => {
  let env: Environment;
  let ctx: Context;
  let factory: Factory;

  const query = gql`
    query User($id: ID!) {
      user(id: $id) {
        id
        role
      }
    }
  `;



  beforeAll(async () => {
    env = await createEnv({ server: true, db: true, cache: true });
  });

  afterAll(async () => {
    await destroyEnv(env);
  });

  beforeEach(async () => {
    ctx = await createContext(env);
    factory = createFactory(ctx);
  });

  afterEach(async () => {
    await destroyContext(ctx);
  });

  it('should return a user', async () => {
    const user = await factory.insert('user', { role: UserRole.Admin });

    ctx.auth.update(user);

    const { data } = await request<Response>(env)
      .withContext(ctx)
      .query(query, { id: user.id })
      .expectNoErrors();
      

    expect(data?.user.id).toEqual(user.id);
    expect(data?.user.role).toEqual(user.role);
  });

  it('should throw error when user does not exist', async () => {
    const user = await factory.insert('user');
    ctx.auth.update(user);
    const { errors } = await request<Response>(env)
      .withContext(ctx)
      .query(query, { id: crypto.randomUUID() });

    expect(errors).toMatchSnapshot();
  });

});

type Response = {
  user: User;
};
