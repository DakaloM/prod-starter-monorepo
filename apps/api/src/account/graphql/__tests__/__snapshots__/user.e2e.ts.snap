// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`account/graphql/user > should throw error when user does not exist 1`] = `
[
  {
    "extensions": {
      "code": "ResourceNotFound",
      "path": [],
    },
    "message": "User not found",
  },
]
`;
